File: core_multiiter_dmdd/hosvd_function.py
import numpy as np
from typing import List, Tuple
from tensor_utils import hosvd


def hosvd_function(Tensor: np.ndarray, varepsilon1: float, nn: List[int], 
                   n: List[int], TimePos: int) -> Tuple[np.ndarray, List[np.ndarray], np.ndarray, List[np.ndarray], List[int]]:
    """
    Perform HOSVD decomposition to calculate the reduced temporal matrix hatT.
    
    Parameters:
    -----------
    Tensor : np.ndarray
        Initial data tensor
    varepsilon1 : float
        Tolerance for SVD
    nn : List[int]
        Current number of singular values
    n : List[int]
        Initial tensor dimensions
    TimePos : int
        Position of temporal dimension (0-indexed)
        
    Returns:
    --------
    tuple
        hatT (reduced temporal matrix), U (temporal SVD modes), 
        S (tensor core), sv (singular values), nn1 (updated dimensions)
    """
    # Perform HOSVD and retain all singular values
    TT, S, U, sv, n_updated = hosvd(Tensor, n)
    
    # Set truncation of singular values using varepsilon1
    nn1 = nn.copy()
    
    for i in range(1, len(nn1)):
        count = 0
        if len(sv[i]) > 0:
            for j in range(len(sv[i])):
                if sv[i][j] / sv[i][0] >= varepsilon1:
                    count += 1
                else:
                    break
        nn1[i] = count
    
    # HOSVD retaining nn1 singular values
    TT, S, U, sv, nn1 = hosvd(Tensor, nn1)
    
    # Construct reduced temporal matrix
    UT = [Ui.T for Ui in U]
    
    hatT = np.zeros((nn1[TimePos], Tensor.shape[TimePos]))
    for kk in range(nn1[TimePos]):
        hatT[kk, :] = sv[TimePos][kk] * UT[TimePos][kk, :]
    
    return hatT, U, S, sv, nn1

File: MADONNE/hosvd_function.m
function [hatT,U,S,sv,nn]=hosvd_function(Tensor,varepsilon1,nn,n,TimePos)

% PERFORM HOSVD AND RETAINS ALL THE SINGULAR VALUES
% [L,I,J,K]=size(Tensor)
% n=[L,I,J,K];
[TT S U sv n] = hosvd(Tensor, n);

% SET THE TRUNCATION OF THE SINGULAR VALUES USING varepsilon1 (automatic
% truncation)
%nn=[L,0,0,0];

for i=2:length(nn)
    count=0;
    for j=1:size(sv{i},1)
        if sv{i}(j)/sv{i}(1)>=varepsilon1
            count=count+1;
        else
            break
        end
    end
    nn(i)=count;
end

('Initial number of singular values')
n
('Number of singular values retained')
nn

% HOSVD retaining n singular values: reconstruction of the mdoes
[TT S U sv nn] = hosvd(Tensor, nn);

% Construct the reduced matrix containing the temporal modes
for pp=1:length(nn)
    UT{pp}=U{pp}';
end

for kk=1:nn(TimePos)
    hatT(kk,:)=sv{TimePos}(kk)*UT{TimePos}(kk,:);
end